File: packages/shared/src/agent/tools/security.js
'use strict';

const fs = require('fs');
const path = require('path');
const { execTool, resolveWorkdir, commandExists, platformCmd, defineTool } = require('./exec');

/**
 * Security tools — run established scanners over the workspace. Everything
 * delegates to a pre-installed scanner (Trivy, npm audit, pip-audit, Semgrep,
 * gitleaks, TruffleHog); we never re-implement detection rules.
 *
 * Secret hygiene: the secret scanner is invoked in REDACT mode (gitleaks
 * `--redact`, TruffleHog `--only-verified`) and all tool output additionally
 * passes through the registry's secret redactor, so found secrets are not
 * echoed verbatim to the model.
 */

/** Pick a dependency/vuln scanner given availability + detected ecosystem. */
async function pickVulnScanner(dir, requested) {
  const has = (f) => fs.existsSync(path.join(dir, f));
  const scanners = {
    trivy: { command: 'trivy', args: ['fs', '--scanners', 'vuln,misconfig,secret', '--exit-code', '0', '.'], hint: 'Install Trivy (https://aquasecurity.github.io/trivy).' },
    'npm-audit': { command: platformCmd('npm'), args: ['audit'], hint: 'Install Node.js (bundles npm).' },
    'pip-audit': { command: 'pip-audit', args: [], hint: 'pip install pip-audit.' },
    semgrep: { command: 'semgrep', args: ['scan', '--config', 'auto', '--error'], hint: 'Install Semgrep (pip install semgrep).' },
  };
  if (requested && requested !== 'auto') return { key: requested, ...scanners[requested] };
  if (await commandExists('trivy')) return { key: 'trivy', ...scanners.trivy };
  if (has('package.json')) return { key: 'npm-audit', ...scanners['npm-audit'] };
  if ((has('requirements.txt') || has('pyproject.toml')) && (await commandExists('pip-audit'))) return { key: 'pip-audit', ...scanners['pip-audit'] };
  if (await commandExists('semgrep')) return { key: 'semgrep', ...scanners.semgrep };
  return { key: 'npm-audit', ...scanners['npm-audit'] }; // last resort; reports "not installed" cleanly
}

const securityScanTool = defineTool(
  {
    name: 'security_scan',
    description:
      'Scan the workspace for dependency vulnerabilities and misconfigurations using the best available scanner ' +
      '(Trivy, npm audit, pip-audit, or Semgrep). Prefer this over ad-hoc grepping for CVEs.',
    schema: (z) =>
      z.object({
        dir: z.string().optional().describe('workspace-relative directory'),
        scanner: z.enum(['auto', 'trivy', 'npm-audit', 'pip-audit', 'semgrep']).optional().describe('force a scanner (default: auto)'),
      }),
  },
  async (input, ctx) => {
    const dir = resolveWorkdir(ctx, input.dir);
    const s = await pickVulnScanner(dir, input.scanner);
    return execTool({ ctx, label: `security scan (${s.key})`, command: s.command, args: s.args, dir: input.dir, notFoundHint: s.hint });
  }
);

const secretScanTool = defineTool(
  {
    name: 'secret_scan',
    description:
      'Scan the workspace for committed secrets/credentials with gitleaks or TruffleHog (run in redacted mode). ' +
      'Use before opening a PR to catch leaked keys.',
    schema: (z) => z.object({ dir: z.string().optional().describe('workspace-relative directory') }),
  },
  async (input, ctx) => {
    if (await commandExists('gitleaks', 'version')) {
      return execTool({ ctx, label: 'secret scan (gitleaks)', command: 'gitleaks', args: ['detect', '--no-banner', '--redact', '--source', '.'], dir: input.dir });
    }
    return execTool({
      ctx,
      label: 'secret scan (trufflehog)',
      command: 'trufflehog',
      args: ['filesystem', '.', '--only-verified', '--no-update'],
      dir: input.dir,
      notFoundHint: 'Install gitleaks (preferred) or TruffleHog to scan for committed secrets.',
    });
  }
);

const FACTORIES = Object.freeze({ security_scan: securityScanTool, secret_scan: secretScanTool });

module.exports = { FACTORIES, pickVulnScanner };

File: ai_fleet/agent/tools/security.py
"""Security tools (port of agent/tools/security.js).

Run established scanners over the workspace. Everything delegates to a
pre-installed scanner (Trivy, npm audit, pip-audit, Semgrep, gitleaks,
TruffleHog); we never re-implement detection rules.

Secret hygiene: the secret scanner is invoked in REDACT mode (gitleaks
``--redact``, TruffleHog ``--only-verified``) and all tool output additionally
passes through the registry's secret redactor, so found secrets are not echoed
verbatim to the model.
"""

from __future__ import annotations

import os
from typing import Literal, Optional

from pydantic import BaseModel, Field

from ai_fleet.agent.tools.exec import (
    command_exists,
    define_tool,
    exec_tool,
    platform_cmd,
    resolve_workdir,
)


async def pick_vuln_scanner(dir, requested=None):
    """Pick a dependency/vuln scanner given availability + detected ecosystem."""
    def has(f):
        return os.path.exists(os.path.join(dir, f))

    scanners = {
        "trivy": {
            "command": "trivy",
            "args": ["fs", "--scanners", "vuln,misconfig,secret", "--exit-code", "0", "."],
            "hint": "Install Trivy (https://aquasecurity.github.io/trivy).",
        },
        "npm-audit": {"command": platform_cmd("npm"), "args": ["audit"], "hint": "Install Node.js (bundles npm)."},
        "pip-audit": {"command": "pip-audit", "args": [], "hint": "pip install pip-audit."},
        "semgrep": {"command": "semgrep", "args": ["scan", "--config", "auto", "--error"], "hint": "Install Semgrep (pip install semgrep)."},
    }
    if requested and requested != "auto":
        return {"key": requested, **scanners[requested]}
    if await command_exists("trivy"):
        return {"key": "trivy", **scanners["trivy"]}
    if has("package.json"):
        return {"key": "npm-audit", **scanners["npm-audit"]}
    if (has("requirements.txt") or has("pyproject.toml")) and (await command_exists("pip-audit")):
        return {"key": "pip-audit", **scanners["pip-audit"]}
    if await command_exists("semgrep"):
        return {"key": "semgrep", **scanners["semgrep"]}
    return {"key": "npm-audit", **scanners["npm-audit"]}  # last resort; reports "not installed" cleanly


class _SecurityScanSchema(BaseModel):
    dir: Optional[str] = Field(default=None, description="workspace-relative directory")
    scanner: Optional[Literal["auto", "trivy", "npm-audit", "pip-audit", "semgrep"]] = Field(
        default=None, description="force a scanner (default: auto)"
    )


class _SecretScanSchema(BaseModel):
    dir: Optional[str] = Field(default=None, description="workspace-relative directory")


async def _security_scan(input, ctx):
    dir_ = resolve_workdir(ctx, input.get("dir"))
    scanner = await pick_vuln_scanner(dir_, input.get("scanner"))
    return await exec_tool(
        ctx=ctx,
        label=f"security scan ({scanner['key']})",
        command=scanner["command"],
        args=scanner["args"],
        dir=input.get("dir"),
        not_found_hint=scanner["hint"],
    )


async def _secret_scan(input, ctx):
    if await command_exists("gitleaks", "version"):
        return await exec_tool(
            ctx=ctx,
            label="secret scan (gitleaks)",
            command="gitleaks",
            args=["detect", "--no-banner", "--redact", "--source", "."],
            dir=input.get("dir"),
        )
    return await exec_tool(
        ctx=ctx,
        label="secret scan (trufflehog)",
        command="trufflehog",
        args=["filesystem", ".", "--only-verified", "--no-update"],
        dir=input.get("dir"),
        not_found_hint="Install gitleaks (preferred) or TruffleHog to scan for committed secrets.",
    )


security_scan_tool = define_tool(
    {
        "name": "security_scan",
        "description": (
            "Scan the workspace for dependency vulnerabilities and misconfigurations using the best available scanner "
            "(Trivy, npm audit, pip-audit, or Semgrep). Prefer this over ad-hoc grepping for CVEs."
        ),
        "schema": _SecurityScanSchema,
    },
    _security_scan,
)

secret_scan_tool = define_tool(
    {
        "name": "secret_scan",
        "description": (
            "Scan the workspace for committed secrets/credentials with gitleaks or TruffleHog (run in redacted mode). "
            "Use before opening a PR to catch leaked keys."
        ),
        "schema": _SecretScanSchema,
    },
    _secret_scan,
)

FACTORIES = {"security_scan": security_scan_tool, "secret_scan": secret_scan_tool}
